src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets

```diff
@@ -40,4 +41,93 @@
       <UseSharedCompilation>true</UseSharedCompilation>
     </PropertyGroup>
   </Target>
+
+  <!--
+    ========================
+    DeterministicSourcePaths
+    ========================
+    
+    Unless specified otherwise enable deterministic source root (PathMap) when building deterministically on CI server, but not for local builds.
+    In order for the debugger to find source files when debugging a locally built binary the PDB must contain original, unmapped local paths.
+  -->
+  <PropertyGroup>
+    <DeterministicSourcePaths Condition="'$(DeterministicSourcePaths)' == '' and '$(Deterministic)' == 'true' and '$(ContinuousIntegrationBuild)' == 'true'">true</DeterministicSourcePaths>
+  </PropertyGroup>
+
+  <!--
+    ==========
+    SourceRoot
+    ==========
+
+    All source files of the project are expected to be located under one of the directories specified by SourceRoot item group.
+    This targets collects all SourceRoots from various sources.
+
+    This target calculates final local path for each SourceRoot and sets SourceRoot.MappedPath metadata accordingly.
+    The final path is a path when DeterministicSourcePaths is true, and the original path otherwise.
+    In addition, the target validates and deduplicates the SourceRoot items.
+
+    InitializeSourceControlInformation is an msbuild target that ensures the SourceRoot items are populated from source control.
+    The target is available only if SourceControlInformationFeatureSupported is true.
+
+    A consumer of SourceRoot.MappedPath metadata, such as Source Link generator, shall depend on this target. 
+  -->
+
+  <Target Name="InitializeSourceRootMappedPaths"
+          DependsOnTargets="_InitializeSourceRootMappedPathsFromSourceControl">
+
+    <ItemGroup Condition="'@(_MappedSourceRoot)' != ''">
+      <_MappedSourceRoot Remove="@(_MappedSourceRoot)" />
+    </ItemGroup>
+
+    <Microsoft.CodeAnalysis.BuildTasks.MapSourceRoots SourceRoots="@(SourceRoot)" Deterministic="$(DeterministicSourcePaths)">
+      <Output TaskParameter="MappedSourceRoots" ItemName="_MappedSourceRoot" />
+    </Microsoft.CodeAnalysis.BuildTasks.MapSourceRoots>
+
+    <ItemGroup>
+      <SourceRoot Remove="@(SourceRoot)" />
+      <SourceRoot Include="@(_MappedSourceRoot)" />
```

Comment:
Re-using the `SourceRoot` item group seems fishy. Why not create a new one (`MappedSourceRoot`) so there is no confusion?

Reply:
This is a common msbuild pattern: tasks add extra metadata to existing item groups. There is only one concept - SourceRoot. Various targets in the build system can add extra metadata on its items.